### PR TITLE
Add bank aggregator service

### PR DESCRIPTION
**Pull request by randrii**

### CODE REVIEW

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import org.springframework.stereotype.Service;
+
+import java.text.SimpleDateFormat;
+import java.util.List;
+
+@Service
+@Data
+public class BankAggregatorService {
+
+    @SneakyThrows
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: set try block and re-throw a custom exception.

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    public void loadExchangeRatesFor(String currency) {
+        var bankExchangePage = Jsoup.connect(String.format(BankAggregatorConstants.BANK_AGGREGATOR_URL, currency)).get();
+        var bankDataList = parseExchangeRatesPage(bankExchangePage);
+    }
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: return bankData? 
what's sense of void here?

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        var bankLinkList = exchangeTableBody.select(BankAggregatorConstants.BANK_LINK)
+                .select(BankAggregatorConstants.BANK_LINK_ATTRIBUTE).eachAttr(BankAggregatorConstants.BANK_LINK_HREF);
+        var bankBuyList = exchangeTableBody.select(BankAggregatorConstants.BANK_BUY_RATE).eachText();
+        var bankSellList = exchangeTableBody.select(BankAggregatorConstants.BANK_SELL_RATE).eachText();
+        var bankUpdateDateList = exchangeTableBody.select(BankAggregatorConstants.BANK_LAST_UPDATE_DATE).eachText();
+        var formatter = new SimpleDateFormat(BankAggregatorConstants.DATE_PATTERN);
+
+        List<BankData> bankDataList = new java.util.ArrayList<>();
+
+        for (int i = 0; i < bankHeaderList.size(); i++) {
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: create separate method

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+@Data
+public class BankAggregatorService {
+
+    @SneakyThrows
+    public void loadExchangeRatesFor(String currency) {
+        var bankExchangePage = Jsoup.connect(String.format(BankAggregatorConstants.BANK_AGGREGATOR_URL, currency)).get();
+        var bankDataList = parseExchangeRatesPage(bankExchangePage);
+    }
+
+    @SneakyThrows
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: remove @SneakyThrows and don't use it in the production code(only tests)

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    private List<BankData> parseExchangeRatesPage(Document document) {
+        var exchangeTableBody = document.getElementsByClass(BankAggregatorConstants.AGGREGATOR_TABLE_BODY);
+        var bankHeaderList = exchangeTableBody.select(BankAggregatorConstants.BANK_HEADER).eachText();
+        var bankLinkList = exchangeTableBody.select(BankAggregatorConstants.BANK_LINK)
+                .select(BankAggregatorConstants.BANK_LINK_ATTRIBUTE).eachAttr(BankAggregatorConstants.BANK_LINK_HREF);
+        var bankBuyList = exchangeTableBody.select(BankAggregatorConstants.BANK_BUY_RATE).eachText();
+        var bankSellList = exchangeTableBody.select(BankAggregatorConstants.BANK_SELL_RATE).eachText();
+        var bankUpdateDateList = exchangeTableBody.select(BankAggregatorConstants.BANK_LAST_UPDATE_DATE).eachText();
+        var formatter = new SimpleDateFormat(BankAggregatorConstants.DATE_PATTERN);
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: Formatter can be created in class field.

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java (hunk 2 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: fixed

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java (hunk 5 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: fixed

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java (hunk 4 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: fixed

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java (hunk 1 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: fixed

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java (hunk 3 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: fixed

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    private final HttpClient httpClient;
+    private final String DATE_PATTERN = "yyyy-M-dd hh:mm";
+
+    public List<BankData> loadExchangeRatesPageFor(String currency) {
+        try {
+            var request = HttpRequest.newBuilder()
+                    .uri(URI.create((String.format(BankAggregatorConstants.BANK_AGGREGATOR_URL, currency))))
+                    .build();
+            var response = httpClient.send(request,
+                    HttpResponse.BodyHandlers.ofString());
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: add separate method for request and response.

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+@Data
+@RequiredArgsConstructor
+public class BankAggregatorService {
+
+    private final HttpClient httpClient;
+    private final String DATE_PATTERN = "yyyy-M-dd hh:mm";
+
+    public List<BankData> loadExchangeRatesPageFor(String currency) {
+        try {
+            var request = HttpRequest.newBuilder()
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: You can use CvURL to reduce code.

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java (hunk 6 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: fixed

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java (hunk 7 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: fixed

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+                .select(BankAggregatorConstants.BANK_LINK_ATTRIBUTE).attr(BankAggregatorConstants.BANK_LINK_HREF);
+        var bankBuyRate = trElement.select(BankAggregatorConstants.BANK_BUY_RATE).text();
+        var bankSellRate = trElement.select(BankAggregatorConstants.BANK_SELL_RATE).text();
+        var bankLastUpdateDate = trElement.select(BankAggregatorConstants.BANK_LAST_UPDATE_DATE).text();
+
+        return buildObject(bankName, bankLink, bankBuyRate, bankSellRate, bankLastUpdateDate);
+    }
+
+    private BankData buildObject(String bankName, String bankLink, String bankBuyRate, String bankSellRate, String bankLastUpdateDate) {
+        var formatter = new SimpleDateFormat(DATE_PATTERN);
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: can be on the class level.

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+import java.util.stream.Collectors;
+
+@Service
+@Slf4j
+@Data
+@RequiredArgsConstructor
+public class BankAggregatorService {
+
+    private final CVurl cVurl;
+    private final String DATE_PATTERN = "yyyy-M-dd hh:mm";
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: Why final? 
Seems spring will try to inject this value.

Comment (jarc0der) [reply src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: We don't need provide pattern in the constructor.;

File: src/test/java/com/rybka/service/exchange/BankAggregatorServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        when(cVurl.get((String.format(BankAggregatorConstants.BANK_AGGREGATOR_URL, targetCurrency)))).thenReturn(requestBuilder);
+        when(requestBuilder.asString()).thenReturn(stringResponse);
+        when(stringResponse.get()).thenReturn(response);
+        when(response.getBody()).thenReturn(bankAggregatorContent);
+
+        // when
+        var bankDataList = service.loadExchangeRatesPageFor(targetCurrency);
+
+        // then
+        Assertions.assertEquals(expectedBankDataListSize, bankDataList.size());
```
Comment (jarc0der) [inline src/test/java/com/rybka/service/exchange/BankAggregatorServiceTest.java]: Verify count is good but on enough. 
We are interested in the content of the parsed data and we want to be sure that parser works and retrieves all expected data.

Comment (jarc0der) [reply src/test/java/com/rybka/service/exchange/BankAggregatorServiceTest.java]: You can left only 3 rows in the table and create list of expected objects.

File: src/main/java/com/rybka/model/BankData.java [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+import java.util.Date;
+
+@Data
+@AllArgsConstructor
+public class BankData {
+    private String name;
+    private String link;
+    private double buy;
+    private double sell;
+    private Date lastUpdated;
```
Comment (jarc0der) [inline src/main/java/com/rybka/model/BankData.java]: Use LocalDateTime it's far more convenient to work with it then Date object.

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+        var bankSellRate = trElement.select(BankAggregatorConstants.BANK_SELL_RATE).text();
+        var bankLastUpdateDate = trElement.select(BankAggregatorConstants.BANK_LAST_UPDATE_DATE).text();
+
+        return buildObject(bankName, bankLink, bankBuyRate, bankSellRate, bankLastUpdateDate);
+    }
+
+    private BankData buildObject(String bankName, String bankLink, String bankBuyRate, String bankSellRate, String bankLastUpdateDate) {
+        var formatter = new SimpleDateFormat(DATE_PATTERN);
+
+        try {
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: This try block can only catch NPE or InvalidArgumentException from Double class. 

Wrap code in the loadExchangeRatesPageFor method into try catch block. That allow us catch all exceptions.

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        var document = Jsoup.parse(retrieveResourcePage(currency).getBody());
+        var trElementRowList = document.select(BankAggregatorConstants.BANK_TABLE_ROW);
+
+        return trElementRowList.stream().map(this::parseBankDataRow).collect(Collectors.toList());
+    }
+
+    private Response<String> retrieveResourcePage(String currency) {
+        return cVurl
+                .get((String.format(BankAggregatorConstants.BANK_AGGREGATOR_URL, currency)))
+                .asString().get();
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: or throw exception?

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+        return buildObject(bankName, bankLink, bankBuyRate, bankSellRate, bankLastUpdateDate);
+    }
+
+    private BankData buildObject(String bankName, String bankLink, String bankBuyRate, String bankSellRate, String bankLastUpdateDate) {
+        var formatter = new SimpleDateFormat(DATE_PATTERN);
+
+        try {
+            return new BankData(bankName, bankLink, Double.parseDouble(bankBuyRate), Double.parseDouble(bankSellRate),
+                    formatter.parse(bankLastUpdateDate.replace('.', '-')));
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: Create constants for `.` and `-` chars.

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+        var bankLastUpdateDate = trElement.select(BankAggregatorConstants.BANK_LAST_UPDATE_DATE).text();
+
+        return buildObject(bankName, bankLink, bankBuyRate, bankSellRate, bankLastUpdateDate);
+    }
+
+    private BankData buildObject(String bankName, String bankLink, String bankBuyRate, String bankSellRate, String bankLastUpdateDate) {
+        var formatter = new SimpleDateFormat(DATE_PATTERN);
+
+        try {
+            return new BankData(bankName, bankLink, Double.parseDouble(bankBuyRate), Double.parseDouble(bankSellRate),
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: I'd do parsing double values in the parseBankDataRow. 
for ex 
var bankBuyRate = Double.parseDouble(trElement.select(BankAggregatorConstants.BANK_BUY_RATE).text())

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java (hunk 8 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: fixed

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java (hunk 9 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: fixed

File: src/main/java/com/rybka/model/BankData.java (hunk 11 above)
Comment (randrii) [reply src/main/java/com/rybka/model/BankData.java]: fixed

File: src/test/java/com/rybka/service/exchange/BankAggregatorServiceTest.java (hunk 10 above)
Comment (randrii) [reply src/test/java/com/rybka/service/exchange/BankAggregatorServiceTest.java]: fixed

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java (hunk 14 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: fixed

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java (hunk 15 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: fixed

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java (hunk 13 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: fixed

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java (hunk 12 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: fixed

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+    private BankData parseBankDataRow(Element trElement) {
+        var bankName = trElement.select(BankAggregatorConstants.BANK_HEADER).text();
+        var bankLink = trElement.select(BankAggregatorConstants.BANK_LINK)
+                .select(BankAggregatorConstants.BANK_LINK_ATTRIBUTE).attr(BankAggregatorConstants.BANK_LINK_HREF);
+        var bankBuyRate = Double.parseDouble(trElement.select(BankAggregatorConstants.BANK_BUY_RATE).text());
+        var bankSellRate = Double.parseDouble(trElement.select(BankAggregatorConstants.BANK_SELL_RATE).text());
+        var bankLastUpdateDate = LocalDateTime.parse(trElement.select(BankAggregatorConstants.BANK_LAST_UPDATE_DATE).text()
+                .replace(BankAggregatorConstants.BANK_DATE_SEPARATOR_OLD, BankAggregatorConstants.BANK_DATE_SEPARATOR_NEW), formatter);
+
+        return buildObject(bankName, bankLink, bankBuyRate, bankSellRate, bankLastUpdateDate);
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: Could be new BankData(bankName, bankLink, bankBuyRate, bankSellRate, bankLastUpdateDate);
no need to create a separate method with the same line of code...

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+    private DateTimeFormatter formatter = DateTimeFormatter.ofPattern(DATE_PATTERN);
+
+    public List<BankData> loadExchangeRatesPageFor(String currency) {
+        try {
+            var document = Jsoup.parse(retrieveResourcePage(currency).getBody());
+            var trElementRowList = document.select(BankAggregatorConstants.BANK_TABLE_ROW);
+
+            return trElementRowList.stream().map(this::parseBankDataRow).collect(Collectors.toList());
+        } catch (Exception exception) {
+            log.error(Messages.RESOURCE_PROCESSING_EXCEPTION_MSG);
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: You message will looks like
"An error occur while processing the resource. Reason: "? The reason is empty.

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+            throw new ResourceProcessingException(Messages.RESOURCE_PROCESSING_EXCEPTION_MSG);
+        }
+    }
+
+    private Response<String> retrieveResourcePage(String currency) {
+        return Optional.of(cVurl
+                .get((String.format(BankAggregatorConstants.BANK_AGGREGATOR_URL, currency)))
+                .asString().get()).orElseThrow(() -> new InvalidResponseException(Messages.RESPONSE_EXCEPTION_MSG));
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: Seems we have method getOrThrow or smth like that. Take a look please.

.asString()
.getOrThrow()

Comment (jarc0der) [reply src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: No need wrap into optional.

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java (hunk 16 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: fixed

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java (hunk 17 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: fixed

File: src/main/java/com/rybka/service/exchange/BankAggregatorService.java (hunk 18 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/BankAggregatorService.java]: fixed

**Pull request merged**